run_tests.py

```diff
@@ -1,74 +1,208 @@
+            print(f'{bcolors.FAIL}Outside of acceptable range. \n \t cosh({i}) = {calc}. \n \t actual = {result}. \n \t error of: {error}. \n')
+            return False
+        elif error < 0 and error < PRECISION:
+            print(f'{bcolors.FAIL}Outside of acceptable range. \n \t cosh({i}) = {calc}. \n \t actual = {result}. \n \t error of: {error}. \n')
             return False
     return True
 
+
 # power function test for integer cases
 def pf_test_integer():
+    print(f'{bcolors.HEADER}pf_integer test starting...')
     for i in range(-10, 10):
         for j in range(-10, 10):
             if not (i == 0 and j < 0):
-                calc = power_function([i, j])
-                result = math.pow(i, j)
+                calc = round(power_function([i, j]), 10)
+                result = round(math.pow(i, j), 10)
                 error = calc - result
-                if error > PRECISION:
-                    print(f'Outside of acceptable range: power_function({i},{j}):{calc}, actual: {result}')
+                if error > 0 and error > PRECISION: 
+                    print(f'{bcolors.FAIL}Outside of acceptable range. \n \t power_function({i},{j}) = {calc}. \n \t actual: {result}. \n \t error of: {error} \n')
+                    return False
+                elif error < 0 and error < PRECISION:
+                    print(f'{bcolors.FAIL}Outside of acceptable range. \n \t power_function({i},{j}) = {calc}. \n \t actual: {result}. \n \t error of: {error} \n')
                     return False
     return True
 
 # power function test for decimal cases
 def pf_test_decimal():
+    print(f'{bcolors.HEADER}pf_decimal test starting...')
     for i in range(0, 10):
         for j in range(-10, 10):
             if not (i == 0 and j < 0):
                 i += 0.5
                 j += 0.5
 
-                calc = power_function([i, j])
-                result = math.pow(i, j)
+                calc = round(power_function([i, j]), 3)
+                result = round(math.pow(i, j), 3)
                 error = calc - result
-                if error > PRECISION:
-                    print(f'Outside of acceptable range: power_function:{calc}, actual: {result}')
+                if error > 0 and error > PRECISION:
+                    print(f'{bcolors.FAIL}Outside of acceptable range. \n \t power_function({i},{j}):{calc}. \n \t actual: {result}. \n \t error of: {error} \n')
+                    return False
+                elif error < 0 and error < PRECISION:
+                    print(f'{bcolors.FAIL}Outside of acceptable range. \n \t power_function({i},{j}):{calc}. \n \t actual: {result}. \n \t error of: {error} \n')
                     return False
     return True
 
+# mean absolute deviation test for integer cases
+def mad_test_integer():
+    print(f'{bcolors.HEADER}mad_integer test starting...')
+    for i in range(0, 10): 
+        list = []
+        for j in range(0, 10):
+            x = random.randint(0, 100)
```

Comment:
I've never seen unit tests use random inputs, but then again, I haven't seem many unit tests. 

Do you know if it's a good idea to do this? It'd make things a little unreproducable if we encountered a strange edge case?

Reply:
Honestly, I think we should redo all tests and instead of doing random numbers, focus on 1-2 normal tests and then only do edge cases. Having ten of these tests is useless, but good enough for iteration 1 IMO.